File: platform/console/ui/src/api/blocks.ts
import type { Block } from './types'

/**
 * The block grammar, parsed HERE.
 *
 * The manager passes an agent's text through untouched, exactly as it passes
 * markdown through untouched, and each adapter reads the grammar and renders it
 * to what its surface has. See design D1 in the structured-agent-output change.
 *
 * PARSING IN THE SURFACE IS WHAT MAKES HISTORY WORK. The tags live in
 * `status.runs[].result`, so a transcript rebuilt after a restart parses the
 * same characters a live message carried and renders identically. While the
 * manager parsed, a parsed structure travelled on the live message only and
 * every reopened conversation went flat.
 *
 * THIS FILE IS DUPLICATED, in Go, in channels/telegram/blocks.go. That is the
 * accepted cost of D1. The RECOGNITION RULES are stated once, in the capability
 * spec, and both implementations are written against them: change one, change
 * both, and keep the adversarial table in step.
 */

const TAG_LINE = /^<(\/?)([a-zA-Z][a-zA-Z0-9_-]*)>[ \t]*$/

const TITLE = 'title'
const DETAILS = 'details'

interface Tag {
  name: string
  closing: boolean
}

/**
 * parsedTag reports whether a line is a standalone block tag.
 *
 * No leading whitespace is allowed and trailing whitespace is: indentation means
 * the line is probably inside something, while trailing spaces are invisible and
 * models emit them constantly. Rejecting a tag over a character nobody can see
 * reads as a bug.
 */
function parsedTag(line: string): Tag | null {
  const m = TAG_LINE.exec(line)
  return m ? { closing: m[1] === '/', name: m[2] } : null
}

/**
 * fencedLines marks every line inside a ``` fence, INCLUDING the fence lines.
 * A fenced block is a machine document quoted verbatim, and a sample showing
 * `<details>` must survive being about this grammar.
 */
function fencedLines(lines: string[]): boolean[] {
  const out: boolean[] = []
  let open = false
  for (const l of lines) {
    if (l.trimStart().startsWith('```')) {
      out.push(true)
      open = !open
      continue
    }
    out.push(open)
  }
  return out
}

/** hasCloser reports whether a matching close tag appears later, outside fences. */
function hasCloser(lines: string[], fenced: boolean[], from: number, name: string): boolean {
  for (let i = from; i < lines.length; i++) {
    if (fenced[i]) continue
    const t = parsedTag(lines[i])
    if (t?.closing && t.name.toLowerCase() === name.toLowerCase()) return true
  }
  return false
}

/** oneLine collapses a title. A heading that wraps to three lines is not one. */
function oneLine(s: string): string {
  return s.split(/\s+/).filter(Boolean).join(' ')
}

/**
 * parse turns an agent's reported output into blocks. It is TOTAL: every input
 * yields blocks, and no input loses a character.
 *
 * A tag is recognized only when it stands alone on its own line at line start,
 * forms a well-formed open/close pair, and sits outside fenced code. Anything
 * else is literal text — which is what keeps an OPEN vocabulary safe, because
 * agent output is full of `<` in shell redirects, generics and code.
 */
export function parse(input: string): Block[] {
  const lines = input.split('\n')
  const fenced = fencedLines(lines)

  const out: Block[] = []
  let prose: string[] = []
  let region: string[] = []
  let curName = ''
  let inRegion = false

  const flushProse = () => {
    const t = prose.join('\n').trim()
    if (t) out.push({ role: 'section', text: t })
    prose = []
  }
  const closeRegion = () => {
    const text = region.join('\n').trim()
    const lower = curName.toLowerCase()
    if (lower === TITLE) out.push({ role: 'title', text: oneLine(text) })
    else if (lower === DETAILS) out.push({ role: 'details', text })
    else out.push({ role: 'section', label: curName, text })
    region = []
    curName = ''
    inRegion = false
  }

  lines.forEach((line, i) => {
    const tag = fenced[i] ? null : parsedTag(line)
    if (!tag) {
      ;(inRegion ? region : prose).push(line)
      return
    }
    if (inRegion && tag.closing && tag.name.toLowerCase() === curName.toLowerCase()) {
      closeRegion()
      return
    }
    if (inRegion) {
      // A TAG INSIDE AN OPEN REGION IS LITERAL. The model is flat, and a model
      // that forgot a close tag is far commoner than one nesting deliberately.
      region.push(line)
      return
    }
    if (tag.closing) {
      // A close with no open never formed a pair: literal text.
      prose.push(line)
      return
    }
    // An unpaired OPEN runs to end of output rather than being discarded —
    // losing an agent's words to a grammar slip is the worst failure available.
    void hasCloser(lines, fenced, i + 1, tag.name)
    flushProse()
    curName = tag.name
    inRegion = true
  })

  if (inRegion) closeRegion()
  flushProse()

  const normalized = normalize(out)
  if (normalized.length === 0) {
    // Total means total: empty in, one empty block out, so no caller downstream
    // has to branch.
    return [{ role: 'section', text: input.trim() }]
  }
  return normalized
}

/**
 * normalize applies the ordering rules: the title FIRST wherever it was written,
 * at most one, and the fold LAST.
 *
 * Named sections keep the order the agent wrote them in. With an open vocabulary
 * nothing here can tell which section is the conclusion, so reordering them
 * would be a guess — and NOTHING trims them either: a length budget cut a
 * markdown table from its header on a live install.
 */
function normalize(blocks: Block[]): Block[] {
  let title: Block | undefined
  const mid: Block[] = []
  const folded: string[] = []

  for (const b of blocks) {
    if (b.role === 'title') {
      if (!title) {
        title = b
        continue
      }
      // A SECOND TITLE IS A SECTION. Its words are kept under its own name
      // rather than becoming a heading no surface knows where to put.
      mid.push({ role: 'section', label: TITLE, text: b.text })
      continue
    }
    if (b.role === 'details') {
      if (b.text) folded.push(b.text)
      continue
    }
    if (b.text || b.label) mid.push(b)
  }

  const out: Block[] = []
  if (title) out.push(title)
  out.push(...mid)
  if (folded.length > 0) {
    // ONE FOLD. "The fold" is singular on every surface, and two disclosure
    // controls in one message is a presentation nobody asked for.
    out.push({ role: 'details', text: folded.join('\n\n') })
  }
  return out
}

File: channels/telegram/blocks.go
package main

// The block grammar, parsed HERE.
//
// The manager passes an agent's text through untouched, exactly as it passes
// markdown through untouched, and each adapter reads the grammar and renders it
// to what its transport has. See design D1 in the structured-agent-output
// change.
//
// It moved out of the manager because nothing there consumed it — the manager
// parsed, put the result on the wire, and every consumer was a renderer — and
// because the parse has to happen on the READ path too, which a manager
// composing outbound messages never runs.
//
// THIS FILE IS DUPLICATED, in TypeScript, in the console. That is the accepted
// cost of D1. The RECOGNITION RULES are stated once, in the capability spec, and
// both implementations are written against them: change one, change both, and
// keep the adversarial table in step.

import (
	"regexp"
	"strings"
)

// Role is what a block IS. Three values, and the vocabulary of SECTION NAMES is
// open on top of them: every agent has a different job, and a fixed tier set
// makes an ha-operator contort into an investigation report.
type Role string

const (
	// RoleTitle is the one-line heading, rendered FIRST wherever it was written.
	RoleTitle Role = "title"
	// RoleSection is above-the-fold content. Label is the agent's own name for
	// it, or empty for prose that carried no tag.
	RoleSection Role = "section"
	// RoleDetails is THE FOLD: collapsed by default on every surface.
	RoleDetails Role = "details"
)

// Block is one piece of a parsed message.
type Block struct {
	Role Role `json:"role"`
	// Label is the agent's own section name, present on named sections only.
	// Adapters render it GENERICALLY — an adapter carrying knowledge of any
	// particular agent's section names is the coupling this grammar avoids.
	Label string `json:"label,omitempty"`
	// Text is prose in the contract's markdown subset.
	Text string `json:"text"`
}

// reserved tag names, matched case-insensitively so `<Title>` is not a section
// called "Title" sitting where the heading should be.
const (
	tagTitle   = "title"
	tagDetails = "details"
)

// tagLine matches a line that is NOTHING BUT a tag.
//
// No leading whitespace is allowed and trailing whitespace is: indentation
// means the line is probably inside something (a list, a code sample), while
// trailing spaces are invisible and models emit them constantly. Rejecting a
// tag over a character nobody can see is the kind of strictness that reads as a
// bug.
var tagLine = regexp.MustCompile(`^(/?)([a-zA-Z][a-zA-Z0-9_-]*)>[ \t]*$`)

// parsedTag reports whether a line is a standalone block tag, and which.
func parsedTag(line string) (name string, closing bool, ok bool) {
	if !strings.HasPrefix(line, "<") {
		return "", false, false
	}
	m := tagLine.FindStringSubmatch(line[1:])
	if m == nil {
		return "", false, false
	}
	return m[2], m[1] == "/", true
}

// Parse turns an agent's reported output into blocks. It is TOTAL: every input
// yields blocks, and no input loses a character.
//
// Recognition requires all three of:
//
//   - the tag stands alone on its own line, at line start
//   - it forms a well-formed open/close pair
//   - it is outside fenced code
//
// Anything else is literal text, which is what keeps an OPEN vocabulary safe:
// `if x < y`, `Deployment<T>` and a shell redirect all fail the first condition,
// and a fenced block explaining this very grammar fails the third.
func Parse(s string) []Block {
	lines := strings.Split(s, "\n")
	fenced := fencedLines(lines)

	var (
		out      []Block
		prose    []string
		region   []string
		curName  string
		inRegion bool
	)

	// flushProse closes off untagged text as an unlabelled above-fold section.
	flushProse := func() {
		if t := strings.TrimSpace(strings.Join(prose, "\n")); t != "" {
			out = append(out, Block{Role: RoleSection, Text: t})
		}
		prose = nil
	}
	// closeRegion ends the open region, whatever ended it.
	closeRegion := func() {
		text := strings.TrimSpace(strings.Join(region, "\n"))
		switch {
		case strings.EqualFold(curName, tagTitle):
			out = append(out, Block{Role: RoleTitle, Text: oneLine(text)})
		case strings.EqualFold(curName, tagDetails):
			out = append(out, Block{Role: RoleDetails, Text: text})
		default:
			out = append(out, Block{Role: RoleSection, Label: curName, Text: text})
		}
		region, curName, inRegion = nil, "", false
	}

	for i, line := range lines {
		name, closing, isTag := parsedTag(line)
		if fenced[i] || !isTag {
			if inRegion {
				region = append(region, line)
			} else {
				prose = append(prose, line)
			}
			continue
		}
		switch {
		case inRegion && closing && strings.EqualFold(name, curName):
			closeRegion()
		case inRegion:
			// A TAG INSIDE AN OPEN REGION IS LITERAL. The model is flat, and a
			// model that forgot a close tag is far commoner than one nesting
			// deliberately — so the content stays where it was written rather
			// than being re-parented by a guess.
			region = append(region, line)
		case closing:
			// A close with no open never formed a pair: literal text.
			prose = append(prose, line)
		case hasCloser(lines, fenced, i+1, name):
			flushProse()
			curName, inRegion = name, true
		default:
			// UNPAIRED OPEN. The region runs to end of output rather than being
			// discarded — losing an agent's words to a grammar slip is the worst
			// failure available here.
			flushProse()
			curName, inRegion = name, true
		}
	}
	if inRegion {
		closeRegion()
	}
	flushProse()

	out = normalize(out)
	if len(out) == 0 {
		// Total means total: empty in, one empty block out, so every caller
		// downstream can assume at least one block rather than branching.
		out = []Block{{Role: RoleSection, Text: strings.TrimSpace(s)}}
	}
	return out
}

// hasCloser reports whether a matching close tag appears later, outside fences.
// Only TOP-LEVEL closers count, which is the same flatness the parse assumes.
func hasCloser(lines []string, fenced []bool, from int, name string) bool {
	for i := from; i < len(lines); i++ {
		if fenced[i] {
			continue
		}
		n, closing, ok := parsedTag(lines[i])
		if ok && closing && strings.EqualFold(n, name) {
			return true
		}
	}
	return false
}

// fencedLines marks every line inside a ``` fence, INCLUDING the fence lines
// themselves. A fenced block is a machine document quoted verbatim, and a
// sample showing `<details>` must survive being about this grammar.
func fencedLines(lines []string) []bool {
	out := make([]bool, len(lines))
	open := false
	for i, l := range lines {
		if strings.HasPrefix(strings.TrimSpace(l), "```") {
			out[i] = true
			open = !open
			continue
		}
		out[i] = open
	}
	return out
}

// oneLine collapses a title to a single line. A heading that wraps to three
// lines is not a heading, and every surface would have to cut it differently.
func oneLine(s string) string {
	return strings.Join(strings.Fields(s), " ")
}

// normalize applies the ordering rules: the title FIRST wherever it was
// written, at most one, and the fold LAST. Named sections keep the order the
// agent wrote them in — with an open vocabulary the manager cannot tell which
// section is the conclusion, so reordering them would be a guess.
func normalize(in []Block) []Block {
	var (
		title  *Block
		mid    []Block
		folded []string
	)
	for _, b := range in {
		switch b.Role {
		case RoleTitle:
			if title == nil {
				t := b
				title = &t
				continue
			}
			// A SECOND TITLE IS A SECTION. Keeping its text under its own name
			// preserves every word without inventing a second heading no
			// surface knows where to put.
			b.Role, b.Label = RoleSection, tagTitle
			mid = append(mid, b)
		case RoleDetails:
			if b.Text != "" {
				folded = append(folded, b.Text)
			}
		default:
			if b.Text != "" || b.Label != "" {
				mid = append(mid, b)
			}
		}
	}
	var out []Block
	if title != nil {
		out = append(out, *title)
	}
	out = append(out, mid...)
	if len(folded) > 0 {
		// ONE FOLD. Several `<details>` regions are one folded region joined in
		// written order — "the fold" is singular on every surface, and two
		// disclosure controls in one message is a presentation nobody asked for.
		out = append(out, Block{Role: RoleDetails, Text: strings.Join(folded, "\n\n")})
	}
	return out
}
